### PR TITLE
Replace simp-le with lego and support DNS-01 challenge

### LINKED ISSUE #34941: security.acme: simp_le -> dehydrated or certbot or acme.sh or lego
## Issue description

NixOS's ```security.acme``` is using [```simp_le```](https://github.com/zenhack/simp_le) to obtain Let's Encrypt certificates

```simp_le``` is a very basic and almost unmaintained tool with at least the following problems:
1. it does not support ```dns-01``` challenge, only ```http-01```, so it requires a machine with http-server and public IP address
2. it tries to obtain all the certificates on each service (re)start, without checking that a certificate has been recently obtained and does not need to be obtained again. This quickly results in Let's Encrypt  servers ban the IP address due to excessive usage.

I think ```simp_le``` should be replaced with a more sophisticated and popular alternative, for example [```dehydrated```](https://github.com/lukas2511/dehydrated) or [```certbot```](https://github.com/certbot/certbot) or [```acme.sh```](https://github.com/Neilpang/acme.sh)

### PR DESCRIPTION
**Pull request by m1cr0man**
###### Motivation for this change

Continuation of #63613 

Closes #34941, #35168

I have the following things still to do:

- Add a deprecation notice for `security.acme.certs.<name>.plugins` (I don't know how, need help)
- Similarly add some equivalent of `mkChangedOptionModule` to convert `extraDomains` to a list since custom webroots are no longer possible.

It is mostly backwards compatible, with a few caveats.

 - `extraDomains` can no longer have different webroots to the main webroot for the cert.
 - An email address is now mandatory for CA registration. I added a more globally scoped `security.acme.email` to make transitioning easier, and an assertion.
- Accepting terms of service was a topic discussed in the other PR. For now, I have added
`security.acme.acceptTerms` and defaulted it to `true` to avoid breaking user's configs. There may be reason to change this default depending on what reviewers think.

The following other changes were required:
 - Deprecate `security.acme.certs.<name>.plugins`, as this was specific to simp-le
 - Rename `security.acme.validMin` to `validMinDays`, to avoid confusion and errors. Lego requires the TTL to be specified in days. I have added a `mkChangedOptionModule` to cover this.
 - Add options to cover DNS-01 challenge (`dnsProvider`, `credentialsFile`, `dnsPropagationCheck`)
 - A shared state directory is now used (`/var/lib/acme/.lego`) to avoid account creation rate limits and share credentials between certs

I have tested converting my own simp-le generated certs to lego and it worked fine. I also have lego in production generating 28 HTTP validated certs and a DNS-01 cert with a bind DNS server (also on a NixOS box :smile: ). This is a non-breaking change if you haven't used custom webroots or plugins.

Also I have tried to keep the diff as small as possible but that other PR was so old it took me a full day to manually rebase on top of all the more recent changes to the ACME module. I'd be tempted to do some other cleanup to the module but there's good git blame as it is now.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/security/acme.nix [hunk 1]
```diff
@@ -204,27 +229,48 @@ in
...
                 lpath = "acme/${cert}";
+                apath = "/var/lib/${lpath}";
+                spath = "/var/lib/acme/.lego";
                 rights = if data.allowKeysForGroup then "750" else "700";
-                cmdline = [ "-v" "-d" data.domain "--default_root" data.webroot "--valid_min" cfg.validMin ]
-                          ++ optionals (data.email != null) [ "--email" data.email ]
-                          ++ concatMap (p: [ "-f" p ]) data.plugins
-                          ++ concatLists (mapAttrsToList (name: root: [ "-d" (if root == null then name else "${name}:${root}")]) data.extraDomains)
+                email = if data.email == null then cfg.email else data.email;
+                globalOpts = [ "-d" data.domain "--email" email "--path" "." ]
```
Comment (m1cr0man) [inline nixos/modules/security/acme.nix]: A quick note on `--path .`. Somewhere between lego 3.0.2 and lego 3.2.0, the default path changed from `./.lego` to `~/.lego`. I had to add this statement so that lego always runs in the service's `WorkingDirectory`.

**Comment by m1cr0man** [comment]
Ping @aanderse 

Finally, the PR is here.

**Comment by aanderse** [comment]
Thanks @m1cr0man! This is great work. I hope to review and test this over the next few weeks, switching some of my test (and hopefully production) systems over to this.

**Comment by arianvp** [comment]
I will have a ook at this Tuesday. Thanks for the work

File: nixos/modules/security/acme.nix [hunk 2]
```diff
@@ -233,18 +279,36 @@ in
...
-                    StateDirectory = lpath;
+                    StateDirectory = "acme/.lego ${lpath}";
                     StateDirectoryMode = rights;
-                    WorkingDirectory = "/var/lib/${lpath}";
-                    ExecStart = "${pkgs.simp_le}/bin/simp_le ${escapeShellArgs cmdline}";
+                    WorkingDirectory = spath;
+                    # Only try loading the credentialsFile if the dns challenge is enabled
+                    EnvironmentFile = if data.dnsProvider != null then data.credentialsFile else null;
+                    ExecStart = pkgs.writeScript "acme-start" ''
+                      #!${pkgs.runtimeShell} -e
```
Comment (arianvp) [inline nixos/modules/security/acme.nix]: Is it not possible to do what we want in one run of the binary?

Comment (Mic92) [reply nixos/modules/security/acme.nix]: Does not look like it: https://github.com/go-acme/lego/issues/216

Comment (m1cr0man) [reply nixos/modules/security/acme.nix]: In order to enforce the `validMinDays` you have to run renew instead of run, which seems like an oversight in lego honestly. Run creates the account credentials if they don't exist but forces a new cert to be generated. I was tempted to add explicit logic to check for the cert's credentials files but I feel that's too solution specific. This conditional logic is the same as in the other PR.

File: nixos/modules/security/acme.nix [hunk 3]
```diff
@@ -204,27 +229,48 @@ in
...
+                spath = "/var/lib/acme/.lego";
                 rights = if data.allowKeysForGroup then "750" else "700";
-                cmdline = [ "-v" "-d" data.domain "--default_root" data.webroot "--valid_min" cfg.validMin ]
-                          ++ optionals (data.email != null) [ "--email" data.email ]
-                          ++ concatMap (p: [ "-f" p ]) data.plugins
-                          ++ concatLists (mapAttrsToList (name: root: [ "-d" (if root == null then name else "${name}:${root}")]) data.extraDomains)
+                email = if data.email == null then cfg.email else data.email;
+                globalOpts = [ "-d" data.domain "--email" email "--path" "." ]
+                          ++ optionals (cfg.acceptTerms) [ "--accept-tos" ]
+                          ++ optionals (data.dnsProvider != null && !cfg.dnsPropagationCheck) [ "--dns.disable-cp" ]
```
Comment (aanderse) [inline nixos/modules/security/acme.nix]:
```suggestion
                          ++ optionals (data.dnsProvider != null && !data.dnsPropagationCheck) [ "--dns.disable-cp" ]
```

File: nixos/modules/security/acme.nix [hunk 4]
```diff
@@ -173,6 +189,15 @@ in
         '';
       };
 
+      acceptTerms = mkOption {
```
Comment (aanderse) [inline nixos/modules/security/acme.nix]: What do other people think of this? Personally I think it is overkill... but am fairly indifferent I guess :man_shrugging:

**Review by aanderse — CHANGES_REQUESTED** [review]
A couple notes as I work through this...

Some issues with `webroot` at the moment. The `tmpfiles` rules reference `webroot` but for DNS certs this won't be defined.

**Comment by m1cr0man** [comment]
If anyone can help me on my two todo's mentioned in the description that would be great. I tried adding a regular mkRemovedOptionModule under certOpts, with a relative attribute path. In order to change extraDomains I tried mkChangedOptionModule but I doubt this works since the attribute path has not changed, and the warning it would produce wouldn't explain the change to a list.
```diff
   cfg = config.security.acme;
 
   certOpts = { name, ... }: {
+    imports = [
+      (mkRemovedOptionModule [ "plugins" ] ''
+        This option has been removed since simp-le was replaced with LEGo.
+        You may be able to achieve the desired functionality with postRun
+      '')
+      (mkChangedOptionModule [ "extraDomains" ] [ "extraDomains" ] (config: let
+        domains = config.security.acme."${name}".extraDomains;
+      in if isList domains then domains else attrNames domains))
+    ];
+
```

These are the errors I got:
```bash
The option `security.acme.certs.mydomain.warnings' defined in `/root/nixpkgs/nixos/modules/security/acme.nix' does not exist.
# Removing the mkChangedOptionModule....
The option `security.acme.certs.mydomain.assertions' defined in `/root/nixpkgs/nixos/modules/security/acme.nix' does not exist.
```

File: nixos/modules/security/acme.nix [hunk 5]
```diff
@@ -111,6 +96,31 @@ let
...
+
+      dnsProvider = mkOption {
+        type = types.nullOr types.str;
+        default = null;
+        example = "route53";
+        description = "DNS Challenge provider";
+      };
+
+      credentialsFile = mkOption {
+        type = types.str;
```
Comment (aanderse) [inline nixos/modules/security/acme.nix]: @m1cr0man could we possibly make this `types.path` instead please?

**Comment by aanderse** [comment]
> If anyone can help me on my two todo's mentioned in the description that would be great. I tried adding a regular mkRemovedOptionModule under certOpts, with a relative attribute path. In order to change extraDomains I tried mkChangedOptionModule but I doubt this works since the attribute path has not changed, and the warning it would produce wouldn't explain the change to a list.

@m1cr0man this is what @Infinisil had to say on the matter: https://logs.nix.samueldr.com/nixos-dev/2020-01-15#1579087078-1579096064

**Comment by Mic92** [comment]
It also seems that simp-le still uses acmev1 which will be sun setted by letsencrypt in June.

**Comment by arianvp** [comment]
@Mic92 nope don't worry; we already addressed this https://github.com/NixOS/nixpkgs/pull/71953

**Comment by aanderse** [comment]
I'm very happy with the new DNS challenge functionality. After a few minor changes I mentioned it worked flawlessly. Next step is to test migration of existing HTTP challenge based certificates.

Great work @m1cr0man!

**Comment by arianvp** [comment]
@aanderse how do you envision that migration path? for example; would keeping around account information and private keys from simp_le around be a requirement for you?

**Comment by m1cr0man** [comment]
@arianvp Those are already being kept around. I'm not explicitly deleting the JSON files in `/var/lib/acme/${name}` so you can revert back pretty painlessly.

**Comment by aanderse** [comment]
I just tested migration on an existing NixOS which uses this module and it worked very smoothly. I was presented with a clear message indicating I had to specify an e-mail address in my configuration and exactly how to do so :+1: 

Has anyone else tested this yet? Aside from my optional request to make `credentialsFile` a `types.path` instead of `types.str` I think this is ready for review.

**Comment by flokli** [comment]
cc @mweinelt, who expressed some interest in the DNS challenge.

**Comment by m1cr0man** [comment]
> Aside from my optional request to make `credentialsFile` a `types.path` instead of `types.str` I think this is ready for review.

I will make this change I'm just holding out for other reviewers for a while. :)

File: nixos/modules/security/acme.nix [hunk 6]
```diff
@@ -111,6 +96,31 @@ let
...
+        description = ''
+          File containing DNS provider credentials passed as environment variables.
+          See https://go-acme.github.io/lego/dns/ for more information.
```
Comment (mweinelt) [inline nixos/modules/security/acme.nix]: Maybe make the format clearer by mentioning that this is passed into the EnvironmentFile option of the systemd unit.

File: nixos/modules/security/acme.nix [hunk 7]
```diff
@@ -111,6 +96,31 @@ let
...
+        '';
+        example = "/var/src/secrets/example.org-route53-api-token";
+      };
+
+      dnsPropagationCheck = mkOption {
+        type = types.bool;
+        default = true;
+        description = ''
+          Toggles LEGo DNS propagation check, which is used alongside DNS-01
```
Comment (mweinelt) [inline nixos/modules/security/acme.nix]: Legos name just seems to be lowercased, see for example the documentation https://go-acme.github.io/lego/installation/.
```suggestion
          Toggles lego DNS propagation check, which is used alongside DNS-01
```

File: nixos/modules/security/acme.nix [hunk 8]
```diff
@@ -111,6 +96,31 @@ let
...
+          challenge to ensure the DNS entries required are available
```
Comment (mweinelt) [inline nixos/modules/security/acme.nix]: Add a trailing dot to finish the sentence.

File: nixos/modules/security/acme.nix (hunk 4 above)
Comment (mweinelt) [reply nixos/modules/security/acme.nix]: If Let's Encrypt wants users to accept their terms I don't think implicitly accepting it for endusers should be the way to go. This knob is the exact place where we can point users to the terms of services of Lets Encrypt.

This doesn't make much sense though when overriding the API endpoints (`-s|--server`) because then we can't know where the correct ToS are.

File: nixos/modules/security/acme.xml [hunk 9]
```diff
@@ -7,7 +7,7 @@
  <para>
   NixOS supports automatic domain validation &amp; certificate retrieval and
   renewal using the ACME protocol. This is currently only implemented by and
-  for Let's Encrypt. The alternative ACME client <literal>simp_le</literal> is
+  for Let's Encrypt. The alternative ACME client <literal>LEGo</literal> is
```
Comment (mweinelt) [inline nixos/modules/security/acme.xml]: Where did you get that capitalisation from?

**Comment by mweinelt** [comment]
Please expose `--key-type`, as its default might not work for everyone. Lego 3.2.0 looks like this:

>   --key-type value, -k value   Key type to use for private keys. Supported: rsa2048, rsa4096, rsa8192, ec256, ec384. (default: "ec384")

**Comment by m1cr0man** [comment]
Ok I've applied all the requested changes to date. `acceptTerms` now defaults to false but I added an assertion so that it tells you to set it to true and links the LE legal repo. I phrased it in such a way that the LE repo is provided for convenience since it's the default but you should find your own CA's terms.

As for `--key-type` I added an option for that and coded in the default of `ec384` so that we control what type of certs generate and when they change, not lego.

File: nixos/modules/security/acme.nix (hunk 6 above)
Comment (m1cr0man) [reply nixos/modules/security/acme.nix]: How does it sound now?

Comment (mweinelt) [reply nixos/modules/security/acme.nix]: Great, thank you!

**Comment by mweinelt** [comment]
I've successfully generated certificates for my laptop through the RFC2136 plugin against Knot using CNAMES into another zone.

One nit was that the acme-foo.example.com.service unit didn't start directly after rebuild.

<details><summary>Example setup</summary>

### Client

```nix
{
  security.acme = {
    email = "me@example.com";
    acceptTerms = true;
    certs = {
      "foo.example.com" = {
        extraDomains = {
          "foo.bar.example.com" = null;
        };
        keyType = "ec384";
        dnsProvider = "rfc2136";
        credentialsFile = /run/keys/acme-example.com.env;
      };
    };
  };
}
```

```sh
RFC2136_NAMESERVER=ns1.example.com
RFC2136_TSIG_ALGORITHM=hmac-sha256.
RFC2136_TSIG_KEY=foo
RFC2136_TSIG_SECRET="changeme"
LEGO_EXPERIMENTAL_CNAME_SUPPORT=1
```

### Server

Created the following records:
```dns
foo.example.com CNAME foo.example.com._acme.example.com.
foo.bar.example.com CNAME foo.example.com._acme.example.com.
```

And spun up Knot like this:

```nix
{ pkgs, lib, ... }:
let
  zones = {
    "_acme.example.com" = ''
      @	SOA ns1.example.com. devnull.example.com. 00000000 86400 7200 3600000 172800
      @ NS ns1.example.com.
      @ A 192.0.0.1
    '';
  };

  zoneFiles = lib.mapAttrsToList (name: zonedata: (pkgs.writeTextDir "${name}.zone" zonedata)) zones;
  zoneDir = pkgs.buildEnv {
    name = "zonefiles";
    paths = zoneFiles;
  };

in {
  services.knot = {
    enable = true;
    extraArgs = [ "-v" ];
    extraConfig = ''
      server:
        listen: 192.0.0.1@53
        listen: 2001:db8::1@53

      log:
        - target: syslog
          any: debug

      template:
        - id: default
          storage: ${zoneDir}
          dnssec-signing: on
          zonefile-sync: -1
          zonefile-load: difference
          journal-content: changes
          journal-db: /var/lib/knot/journal
          kasp-db: /var/lib/knot/kasp
          timer-db: /var/lib/knot/timer

      key:
        - id: foo
          algorithm: hmac-sha256
          secret: changeme

      acl:
        - id: foo
          key: foo
          action: update
          update-owner: name
          update-owner-match: equal
          update-owner-name: [foo.example.com._acme.example.com]
          update-type: [TXT]

      zone:
        - domain: _acme.example.com
          file: _acme.example.com.zone
          acl: [foo]
    '';
  };
}
```

</details>

**Comment by flokli** [comment]
Very nice writeup, thanks :-)

We should also add a paragraph to the release notes, inside "other notable changes", saying we switched to lego, and that the TOS now need to be accepted.

**Comment by mweinelt** [comment]
And also that DNS Challenges are now possible.

**Comment by aanderse** [comment]
ping (mostly because I'm a huge fan of this PR :smile:)

File: nixos/modules/security/acme.nix (hunk 4 above)
Comment (arianvp) [reply nixos/modules/security/acme.nix]: I don't have a really strong feeling about this. We  were implicitly accepting it before no? This seems to make the behaviour _slightly_ better than before

File: nixos/modules/security/acme.nix [hunk 10]
```diff
@@ -204,27 +229,48 @@ in
...
+          assertion = all (certOpts: certOpts.dnsProvider == null || certOpts.webroot == null) certs;
+          message = ''
+            Options `security.acme.certs.<name>.dnsProvider` and
+            `security.acme.certs.<name>.webroot` are mutually exclusive.
+          '';
+        }
+        {
+          assertion = cfg.email != null || all (certOpts: certOpts.email != null) certs;
+          message = ''
+            You must define `security.acme.certs.<name>.email` or
```
Comment (arianvp) [inline nixos/modules/security/acme.nix]: Is `email` really necessary? I thought it was optional. Only has to be set if you want expiration emails?

File: nixos/modules/security/acme.nix [hunk 11]
```diff
@@ -204,27 +229,48 @@ in
...
+                # StateDirectory must be relative, and will be created under /var/lib by systemd
                 lpath = "acme/${cert}";
+                apath = "/var/lib/${lpath}";
+                spath = "/var/lib/acme/.lego";
                 rights = if data.allowKeysForGroup then "750" else "700";
-                cmdline = [ "-v" "-d" data.domain "--default_root" data.webroot "--valid_min" cfg.validMin ]
-                          ++ optionals (data.email != null) [ "--email" data.email ]
-                          ++ concatMap (p: [ "-f" p ]) data.plugins
-                          ++ concatLists (mapAttrsToList (name: root: [ "-d" (if root == null then name else "${name}:${root}")]) data.extraDomains)
+                email = if data.email == null then cfg.email else data.email;
```
Comment (arianvp) [inline nixos/modules/security/acme.nix]: How about instead do this:


```
    email = mkOption {
        type = types.nullOr types.str;
        default = cfg.email;
        description = "Contact email address for the CA to be able to reach you.";
      };



```

Comment (arianvp) [reply nixos/modules/security/acme.nix]: Then you don't need the `if` expression here.   `data.email` defaults to `cfg.email` (which might be `null` too. See my above question)

**Comment by arianvp** [comment]
Left a few comments. Could you have a look? apart from that I think this is LGTM

**Comment by flokli** [comment]
In addition to the stuff mentioned by @arianvp, We also still need to add a line the release notes.

**Comment by aanderse** [comment]
@m1cr0man please ping @arianvp about his generous offer to help out if you need. We are coming up on feature freeze! https://discourse.nixos.org/t/nixos-20-03-feature-freeze/5655/13

**Comment by m1cr0man** [comment]
Sorry folks I switched jobs recently. Going to try get some of this done now.

File: nixos/modules/security/acme.nix (hunk 10 above)
Comment (m1cr0man) [reply nixos/modules/security/acme.nix]: Just gave it a shot, lego (3.2.0) produced this error:
```
2020/02/03 21:34:22 You have to pass an account (email address) to the program using --email or -m
```

File: nixos/modules/security/acme.nix (hunk 11 above)
Comment (m1cr0man) [reply nixos/modules/security/acme.nix]: Yeah that's way better, implemented :)

**Comment by m1cr0man** [comment]
Made the requested changes to the email stuff. I have learned how to write the test but not had time to actually do it tonight. Do I need to update the changelog? If so how do I do that?

**Comment by m1cr0man** [comment]
Sorry for the triple-comment, but I just want to point out that I never figured out how to add the assertions to the submodule. The `plugins` attribute will raise an assertion anyway since it's not defined anymore. `extraModules` is currently unchanged but setting a custom webroot is now impossible so it makes more sense to convert it to a list. I'm happy to merge without these and just the tests for now, and work on them later.

**Comment by m1cr0man** [comment]
Ok it was super difficult to figure out, but I got the DNS-01 challenge test working. However in doing this I've somehow broken the test for b.example.com and I'm not sure how. Would someone else be able to take a look at it(@arianvp)?

**Comment by m1cr0man** [comment]
Fixed it - http-01 validation was running before nginx was started. :) This is ready for review again

File: nixos/doc/manual/release-notes/rl-2003.xml [hunk 12]
```diff
@@ -441,6 +441,22 @@ users.users.me =
...
+     The ACME module has switched from simp-le to <link xlink:href="https://github.com/go-acme/lego">lego</link>
+     which allows us to support DNS-01 challenges and wildcard certificates. The following options have been added:
+     <link linkend="opt-security.acme.acceptTerms">security.acme.acceptTerms</link>,
+     <link linkend="opt-security.acme.certs">security.acme.certs.&lt;name&gt;.dnsProvider</link>,
+     <link linkend="opt-security.acme.certs">security.acme.certs.&lt;name&gt;.credentialsFile</link>,
+     <link linkend="opt-security.acme.certs">security.acme.certs.&lt;name&gt;.dnsPropagationCheck</link>.
+     As well as this, the options <literal>security.acme.acceptTerms</literal> and either
+     <literal>security.acme.email</literal> or <literal>security.acme.certs.&lt;name&gt;.email</literal>
+     must be set in order to use the ACME module.
+     Certificates will be regenerated from new on the next renewal date. The credentials for simp-le are
```
Comment (mweinelt) [inline nixos/doc/manual/release-notes/rl-2003.xml]:
```suggestion
     Certificates will be regenerated anew on the next renewal date. The credentials for simp-le are
```

File: nixos/tests/acme.nix [hunk 13]
```diff
@@ -1,17 +1,51 @@
...
+      ];
+    };
+
+    dnsserver = { nodes, pkgs, ... }: {
+      networking.firewall.allowedTCPPorts = [ 8055 53 ];
+      networking.firewall.allowedUDPPorts = [ 53 ];
+      systemd.services.pebble-challtestsrv = {
+        enable = true;
+        description = "Pebble ACME challenge test server";
+        requires = [ ];
```
Comment (mweinelt) [inline nixos/tests/acme.nix]: Any reason you have an empty requires here?

File: nixos/tests/acme.nix [hunk 14]
```diff
@@ -79,15 +108,48 @@ in import ./make-test-python.nix {
...
+        ({pkgs, config, nodes, lib, ...}: {
+          security.acme.certs."example.com" = {
+            domain = "*.example.com";
+            dnsProvider = "exec";
+            dnsPropagationCheck = false;
+            credentialsFile = with pkgs; writeText "wildcard.env" ''
+              EXEC_PATH=${dnsScript { inherit writeScript bash curl; dnsAddress = nodes.dnsserver.config.networking.primaryIPAddress; }}
+            '';
+            user = config.services.nginx.user;
+            group = config.services.nginx.group;
```
Comment (flokli) [inline nixos/tests/acme.nix]: Why is this necessary? Can't we use the per-vhost `enableACME` option in the nginx module, and just reconfigure the dns provider in the acme module?

File: nixos/tests/acme.nix [hunk 15]
```diff
@@ -79,15 +108,48 @@ in import ./make-test-python.nix {
...
+          };
+          systemd.targets."acme-finished-example.com" = {};
```
Comment (flokli) [inline nixos/tests/acme.nix]: Why do we do that? This looks like you're workarounding something…

File: nixos/tests/acme.nix [hunk 16]
```diff
@@ -79,15 +108,48 @@ in import ./make-test-python.nix {
...
+          systemd.services."acme-example.com" = {
+            wants = [ "acme-finished-example.com.target" ];
+            before = [ "acme-finished-example.com.target" "nginx.service" ];
+            wantedBy = [ "nginx.service" ];
+          };
```
Comment (flokli) [inline nixos/tests/acme.nix]: Same here - I wouldn't expect to have to specify that.

File: nixos/tests/acme.nix [hunk 17]
```diff
@@ -79,15 +108,48 @@ in import ./make-test-python.nix {
...
+          services.nginx.virtualHosts."c.example.com" = {
+            forceSSL = true;
```
Comment (flokli) [inline nixos/tests/acme.nix]: why can't we use `enableACME` here?

File: nixos/tests/acme.nix [hunk 18]
```diff
@@ -79,15 +108,48 @@ in import ./make-test-python.nix {
...
+            sslCertificate = config.security.acme.certs."example.com".directory + "/cert.pem";
+            sslTrustedCertificate = config.security.acme.certs."example.com".directory + "/full.pem";
+            sslCertificateKey = config.security.acme.certs."example.com".directory + "/key.pem";
```
Comment (flokli) [inline nixos/tests/acme.nix]: This is then probably not necessary anymore, too.

File: nixos/tests/acme.nix (hunk 15 above)
Comment (m1cr0man) [reply nixos/tests/acme.nix]: There's a comment on the earlier ones. It's so that it can be waited upon in the tests.

File: nixos/tests/acme.nix (hunk 14 above)
Comment (m1cr0man) [reply nixos/tests/acme.nix]: I'm testing wildcard certs. You can't specify a wildcard cert through enableACME, and that option is tested elsewhere. Personally I use the cert for more than just nginx so I want to ensure it works hard way.

File: nixos/tests/acme.nix (hunk 16 above)
Comment (m1cr0man) [reply nixos/tests/acme.nix]: nginx was starting after the certs consistently when I was testing, causing the HTTP-01 validation to fail. Roll it back and run `nix-build acme.nix` and you'll see what I mean

File: nixos/tests/acme.nix (hunk 17 above)
Comment (m1cr0man) [reply nixos/tests/acme.nix]: I'm testing wildcard certs. I don't want a cert for c.example.com

File: nixos/tests/acme.nix (hunk 13 above)
Comment (m1cr0man) [reply nixos/tests/acme.nix]: Copied from pebble service. Will drop it.

File: nixos/tests/acme.nix (hunk 15 above)
Comment (flokli) [reply nixos/tests/acme.nix]: Please copy that comment over - I got drowned in the rest.

File: nixos/tests/acme.nix (hunk 14 above)
Comment (flokli) [reply nixos/tests/acme.nix]: I'm not sure - I'd expect `serverAliases` to work with that - if not, I'd consider it a bug.

I just find this test hard to understand. - If we can't use the nginx-provided ACME functionalities, probably it'd be more understandable to just not use nginx in the test, only configure the acme module, and check test certificate files appear.

File: nixos/tests/acme.nix (hunk 16 above)
Comment (flokli) [reply nixos/tests/acme.nix]: That's only for the wildcard certificate, right?

File: nixos/tests/acme.nix [hunk 19]
```diff
@@ -61,14 +93,11 @@ in import ./make-test-python.nix {
...
-            ${config.networking.primaryIPAddress} b.example.com
-          '';
           systemd.targets."acme-finished-b.example.com" = {};
           systemd.services."acme-b.example.com" = {
             wants = [ "acme-finished-b.example.com.target" ];
             before = [ "acme-finished-b.example.com.target" ];
+            after = [ "nginx.service" ];
           };
           services.nginx.virtualHosts."b.example.com" = {
             enableACME = true;
```
Comment (flokli) [inline nixos/tests/acme.nix]: we should name these "subdomain certs" a bit more self-explaining, like "webroot-nginx" maybe?

File: nixos/tests/acme.nix (hunk 17 above)
Comment (flokli) [reply nixos/tests/acme.nix]: see above.

File: nixos/tests/acme.nix (hunk 14 above)
Comment (m1cr0man) [reply nixos/tests/acme.nix]: I don't think it's sufficient to simply check does the file exist to truly test wildcard certs. It should work when used with any service (in this case, nginx) serving under that domain. Using nginx here keeps things consistent with the other tests. You're right I could've jused used `serverAliases` and added `*.example.com` there and it would've worked fine, but I'm testing the acme module, not nginx.

File: nixos/tests/acme.nix (hunk 16 above)
Comment (m1cr0man) [reply nixos/tests/acme.nix]: No it was for the HTTP-01 challenge certs, aka the ones using `enableACME`

File: nixos/tests/acme.nix (hunk 14 above)
Comment (flokli) [reply nixos/tests/acme.nix]: I meant using nginx, but not the acme-specific support of it makes the test much more verbose. Maybe a nginx test with`serverAliases`, and a simple one which just looks at the certificate via `openssl x509 -in full.pem -text -noout | grep DNS`?

**Comment by disassembler** [comment]
@m1cr0man are you going to be able to get this addressed before feature freeze tomorrow?

**Comment by flokli** [comment]
This is mostly me being picky with the tests being a bit too complicated for my taste - the module itself was already tested in "production" environments.

FWIW, this can and should be merged for 20.03 - the tests can be improved later on, too.

**Comment by m1cr0man** [comment]
@flokli in general I found these acme tests to be a bit too complex. I would love to rewrite the whole lot of them and incorporate some Apache based tests too, but I can't do that today. I feel that given that there is at least a working test for DNS-01 and the numerous production instances of this code now running that it is stable for 20.03

**Comment by arianvp** [comment]
Yep LGTM and improve the test suite at a later stage

**Comment by flokli** [comment]
Okay, let's merge this, and improve the tests later on.
